Improve Editor Inspector/Theme item lookup performance

aaronp64 opened the pull request:
Changes to reduce the latency between changing node selection in the editor and seeing the new node reflected in the Inspector tab

- Use `Vector` instead of `List` for `ThemeOwner::get_theme_type_dependencies` and related functions
- Use `Vector` instead of `List` for `ThemeContext::themes`, `set_themes()`, and `get_themes()`
- Add `ClassDB:get_inheritance_chain_nocheck` to get all parent/ancestor classes at once, to avoid repeated `ClassDB` locking overhead
- Update `BIND_THEME_ITEM` macros and `ThemeDB::update_class_instance_items` to use provided `StringNames` for call to `ThemeItemSetter`, instead of creating a new `StringName` in each call

These changes reduce the time taken by `EditorInspector::update_tree` by around 30-35%.  This should noticeably improve the delays/freezes in #93402, though there's still a lot going on when updating the Inspector. Still could be slow for some nodes, and would depend on CPU.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

aaronp64 committed:
Improve Editor Inspector/Theme item lookup performance

Changes to reduce the latency between changing node selection in the editor and seeing the new node reflected in the Inspector tab

- Use Vector instead of List for ThemeOwner::get_theme_type_dependencies and related functions
- Use Vector instead of List for ThemeContext::themes, set_themes(), and get_themes()
- Add ClassDB:get_inheritance_chain_nocheck to get all parent/ancestor classes at once, to avoid repeated ClassDB locking overhead
- Update BIND_THEME_ITEM macros and ThemeDB::update_class_instance_items to use provided StringNames for call to ThemeItemSetter, instead of creating a new StringName in each call

These changes reduce the time taken by EditorInspector::update_tree by around 30-35%

Mickeon commented on scene/theme/theme_db.cpp:
```diff
@@ -198,21 +198,21 @@ Ref<StyleBox> ThemeDB::get_fallback_stylebox() {
-		p_list->push_back(class_name);
-		class_name = ClassDB::get_parent_class_nocheck(class_name);
-	}
+	r_result.reserve(8);
```
How was this `8` chosen? An arbitrary number may spell disaster...

AThousandShips replied:
It's just reserving, it doesn't change anything except worsening performance if the guess is wrong

Mickeon replied:
In that case wouldn't it be better to reserve memory in `ClassDB::get_parent_class_nocheck` itself? I know it sounds odd but ClassDB would "know" better (if someone decided to expand the code). Makes the intent more clear, too.

aaronp64 replied:
8's just an estimate of how many classes might be added, ideally a little higher than needed but it should be fine if it needs to grow past that in some cases.  It's mainly to avoid a bunch of small allocations for 1/2/4 items in the vector.  It's based on how many are expected to be added by `ClassDB`, so agreed on moving it there.

Ideally, it would still be reserved before adding `p_base_type`, which might not be in `ClassDB`, so could be unexpected to always add it in there.  Might shuffle things around a bit to make `ClassDB::get_parent_classes_nocheck` match up better with `ClassDB::is_parent_class` (include the passed in class if it's found), and have `ThemeDB::get_native_type_dependencies` add `p_base_type` when needed (if `ClassDB` doesn't find it).

aaronp64 replied:
Made the changes above and renamed `get_parent_classes_nocheck` to `get_inheritance_chain_nocheck`, the old name felt confusing since the starting class is now included.

Mickeon commented on scene/resources/theme.cpp:
```diff
@@ -1250,14 +1250,12 @@ void Theme::get_type_list(List<StringName> *p_list) const {
 }
 
-void Theme::get_type_dependencies(const StringName &p_base_type, const StringName &p_type_variation, List<StringName> *p_list) {
-	ERR_FAIL_NULL(p_list);
```
Is this basically guaranteed to never happen, hence the removal of this line?

AThousandShips replied:
It's not passed by pointer any more but by reference

Mickeon reviewed (approved):
Could use some testing but the code change is surprisingly straightforward.

JekSun97 commented:
This didn't completely solve my problem, but testing showed a small speed increase, I think it can be combined with 4.3

KoBeWi reviewed (approved):
Doesn't seem to break anything.

Not sure if it's a proper usage of LocalVector.

akien-mga commented:
> Not sure if it's a proper usage of LocalVector.

Indeed, `LocalVector` was designed for _local_ usage, hence the name. `Vector` is the one intended for passing around as method argument like this. Would switching to `Vector` reduce the performance gains from this PR?

aaronp64 commented:
Below are the times I got for `EditorInspector::update_tree` when switching between `Tree` nodes:
```
old: 186ms
PR current: 126ms
PR with Vector: 145ms
```

There were two main reasons I went with `LocalVector` instead of `Vector`.  They can both be worked around when using `Vector` (not reflected in the times above), though it makes it more complex to use, and would still probably be a little slower than using `LocalVector`.

- Can't reserve space in `Vector`.  This slows down `ClassDB::get_inheritance_chain_nocheck` if we just use `push_back`.  It would be possible to `resize` up front, set the items, then `resize` again back down to the final size (we don't know the exact size needed until walking through the classes in `ClassDB`).  We'd also have to check for going past the initial `resize`, and switch to `push_back` at that point (or keep resizing/setting).

- `ThemeContext::get_themes` returns a copy of its `themes` member.  I kept this as a copy with this PR, just using `LocalVector` instead of `List`.  If we can change this to a const reference, it wouldn't matter much which container is used, though that would require knowing that `themes` can't be changed while anything calling `get_themes` is using it.  Most uses of `get_themes` look something like `for (const Ref<Theme> &theme : global_context->get_themes())`.  When using `Vector`, there's nothing to say that the returned `Vector` needs to be `const`, so the loop will use `Vector<>::Iterator` instead of `Vector<>::ConstIterator`.  This calls `ptrw()`, so it still ends up doing a full copy.  I originally tried using `Vector` here instead of `LocalVector` to avoid the copy, but since it copies anyway, and probably had some overhead compared to using `LocalVector`, I switched it.  We could add the extra step of using a const reference each time we call `get_themes` to avoid the full copy.  Alternatively, this case is probably similar enough to just use `Vector` (the `ClassDB::get_inheritance_chain_nocheck` case should be more important for performance)

aaronp64 commented:
Updated to use `Vector`, with a couple changes to make it slightly faster than using `LocalVector`:

- Updated `ClassDB::get_inheritance_chain_nocheck` to scan through the classes to resize the `Vector` once, then go through again to actually set the values
- Updated `ThemeContext::get_themes()` to return `const Vector<Ref<Theme>>`, to avoid accidentally calling `ptrw` when looping through the result.  Calls where we need to modify the `Vector` already call with something like `Vector<Ref<Theme>> themes = global_context->get_themes()`, which gives a non-const copy to work with, like in `FontVariation::_get_base_font_or_default() `

KoBeWi commented:
> Calls where we need to modify the Vector already call with something like Vector<Ref<Theme>> themes = global_context->get_themes(), which gives a non-const copy to work with, like in

Does that mean the vector is always copied? Was it like that before?
The code using non-const Vector only modifies it in editor, so it's less performant for no reason if it makes a copy. I'm not that familiar with Vector internals though.

aaronp64 commented:
Sorry, I started using "copy" to mean two different things - calling the `Vector` copy constructor, and actually copying the items inside the container.  Below is what's happening in each version of the code:

- The original `get_themes` returned a `List`, which would always copy each item from the original `List`, and required a separate allocation for each node being created in the new `List`
- The previous version of this PR returned a `LocalVector`, which would always copy each item from the original `LocalVector`, but could copy everything into one allocated block of memory, so it was still faster than `List`
- The current version of this PR returns a `Vector`, which is basically just a new pointer to the existing items.  The items will only be copied to new memory in the case where the `Vector` is modified.

akien-mga commented:
Thanks!
